components/server/src/raft_engine_switch.rs

```diff
@@ -0,0 +1,418 @@
+                                        .append(region_id, std::mem::take(&mut entries))
+                                        .unwrap();
+                                }
+                                _ => unreachable!("There is only 2 types of keys in raft"),
+                            }
+                            // Avoid long log batch.
+                            if local_size >= BATCH_THRESHOLD {
+                                local_size = 0;
+                                batch
+                                    .append(region_id, std::mem::take(&mut entries))
+                                    .unwrap();
+
+                                let size = new_engine.consume(&mut batch, false).unwrap();
+                                count_size.fetch_add(size, Ordering::Relaxed);
+                            }
+                            Ok(true)
+                        }
+                    }
+                },
+            )
+            .unwrap();
+    }
+    let size = new_engine.consume(&mut batch, false).unwrap();
+    count_size.fetch_add(size, Ordering::Relaxed);
+}
+
+pub fn check_and_dump_raft_engine(config: &TiKvConfig, engine: &RocksEngine, thread_num: usize) {
+    let raft_engine_config = config.raft_engine.config();
+    let raft_engine_path = &raft_engine_config.dir;
+
+    let dirty_raft_engine_path = get_path_for_remove(raft_engine_path);
+    if !RaftLogEngine::exists(raft_engine_path) {
+        remove_tmp_dir(&dirty_raft_engine_path);
+        return;
+    }
+
+    // Clean the target engine if it exists.
+    clear_raft_db(engine).expect("clear_raft_db");
+
+    let src_engine = RaftLogEngine::new(raft_engine_config.clone());
+
+    let count_size = Arc::new(AtomicUsize::new(0));
+    let mut count_region = 0;
+    let mut threads = vec![];
+    let (tx, rx) = unbounded();
+    for _ in 0..thread_num {
+        let src_engine = src_engine.clone();
+        let dst_engine = engine.clone();
+        let count_size = count_size.clone();
+        let rx = rx.clone();
+        let t = std::thread::spawn(move || {
+            run_dump_raft_engine_worker(&rx, &src_engine, &dst_engine, &count_size);
+        });
+        threads.push(t);
+    }
+
+    info!("Start to scan raft log from RaftLogEngine and dump into RocksEngine");
+    let consumed_time = std::time::Instant::now();
+    // Seek all region id from RaftLogEngine and send them to workers.
+    for id in src_engine.raft_groups() {
+        tx.send(id).unwrap();
+        count_region += 1;
+    }
+    drop(tx);
+
+    info!("Scanned all region id and waiting for dump");
+    for t in threads {
+        t.join().unwrap();
+    }
+    engine.sync().unwrap();
+    info!(
+        "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
+        count_region,
+        count_size.load(Ordering::Relaxed),
+        consumed_time.elapsed(),
+    );
+
+    rename_to_tmp_dir(&raft_engine_path, &dirty_raft_engine_path);
```

Comment:
what if raft-engine dir is same as raftdb dir

Reply:
Now it's checked in config.rs.